beard/metrics/text.py

```diff
@@ -0,0 +1,126 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Beard.
+# Copyright (C) 2015 CERN.
+#
+# Beard is a free software; you can redistribute it and/or modify it
+# under the terms of the Revised BSD License; see LICENSE file for
+# more details.
+
+"""String metrics.
+
+.. codeauthor:: Petros Ioannidis <petros.ioannidis91@gmail.com>
+
+"""
+
+from __future__ import division
+import re
+
+
+def find_all(s, pattern):
+    """Find all occurences of the given pattern.
+
+    Parameters
+    ----------
+    :param s: string
+        String to be searched
+
+    :param letter: string
+        Substring we are searching for
+
+    Returns
+    -------
+    :returns: generator
+        A generator that holds the indexes of the patterns
+    """
+    for match in re.finditer(pattern, s):
+        yield match.start()
+
+
+def jaro_matching(s1, s2):
+    """Return the number of matching letters.
+
+    Calculate the number of matching letters
+    based on the method for computing the Jaro distance metric.
+
+    Parameters
+    ----------
+    :param s1: string
+        First string
+
+    :param s2: string
+        Second string
+
+    Returns
+    -------
+    :returns: int
+        The number of matching letters
+    """
+    H = min(len(s1), len(s2))//2
+
+    letters_cache = dict()
+    matches = 0
+    transpositions = 0
+    s1_matching_letter_sequence = []
+    s2_matching_letter_sequence = []
+    s1_matched_positions = []
+    s2_matched_positions = []
+
+    for letter in s1:
+        if letter not in letters_cache:
+            letters_cache[letter] = (tuple(find_all(s1, letter)),
+                                     tuple(find_all(s2, letter)))
+
+    for letter, (s1_occur, s2_occur) in letters_cache.items():
+        for pos_s1 in s1_occur:
+            for pos_s2 in s2_occur:
+                if pos_s1 - H <= pos_s2 <= pos_s1 + H:
+                    if pos_s2 not in s2_matched_positions:
+                        matches += 1
+                        s2_matched_positions.append(pos_s2)
+                    s1_matching_letter_sequence.append((pos_s1, letter))
+                    break
+
+    for letter, (s1_occur, s2_occur) in letters_cache.items():
+        for pos_s2 in s2_occur:
+            for pos_s1 in s1_occur:
+                if pos_s1 - H <= pos_s2 <= pos_s1 + H:
+                    if pos_s1 not in s1_matched_positions:
+                        s1_matched_positions.append(pos_s1)
+                    s2_matching_letter_sequence.append((pos_s2, letter))
+                    break
+
+    s1_matching_letter_sequence.sort()
+    s2_matching_letter_sequence.sort()
```

Comment:
Hmm I feel there is a more straightforward way to compute all these values. I am not not sure the caching you do above is worth it in the end because you need to resort the sequences. I'll check this more closely tomorrow.